flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java

```diff
@@ -127,8 +170,62 @@ boolean isProviderLoaded(String serviceName) {
      * Obtains new tokens in a one-time fashion and leaves it up to the caller to distribute them.
      */
     @Override
-    public void obtainDelegationTokens(Credentials credentials) {
+    public void obtainDelegationTokens(Credentials credentials) throws Exception {
         LOG.info("Obtaining delegation tokens");
+
+        // Delegation tokens can only be obtained if the real user has Kerberos credentials, so
+        // skip creation when those are not available.
+        if (kerberosLoginProvider.isLoginPossible()) {
+            UserGroupInformation freshUGI = kerberosLoginProvider.doLogin();
+            freshUGI.doAs(
+                    (PrivilegedExceptionAction<Void>)
+                            () -> {
+                                obtainDelegationTokensAndGetNextRenewal(credentials);
+                                return null;
+                            });
+            LOG.info("Delegation tokens obtained successfully");
+        } else {
+            LOG.info("Real user has no kerberos credentials so no tokens obtained");
+        }
+    }
+
+    protected long obtainDelegationTokensAndGetNextRenewal(Credentials credentials) {
+        AtomicLong nextRenewal = new AtomicLong(Long.MAX_VALUE);
+
+        delegationTokenProviders
+                .values()
+                .forEach(
```

Comment:
You could use a stream, flatmap -> renewal time, and min to get rid of the AtomicLong based min logic

Reply:
I've tried this in the initial implementation but it was just not working. No idea what was the reason so re-checking...

Reply:
If you paste the code here we can probably spot the problem together.

Reply:
Fixed.